### DISCUSSION TITLE
Bug processing <abbreviated-form> elements to PDF

### DISCUSSION THREAD

**Discussion by gershon-pcas** [dita-ot/dita-ot#3937] [Q&A] [answered]
Hi,

I seem to have found a bug in v3.7.1 PDF processing. The OT is not processing <abbreviated-form> elements correctly. The actual error is that it's not recognizing glossentry topics as such and therefore does not process them. Instead, the OT emits warning messages, such as:

`[pipeline] file:/C:/gershon/debug_glossary/reference_topic.dita:10:44: [DOTX060W][WARN]: Key 'glossary_abs' was used in an abbreviated-form element, but the key is not associated with a glossary entry. Abbreviated-form should ONLY be used to reference to a glossary entry.`

The <abbreviated-form> element is dropped from the PDF output, in other words, all references to the glossentry term are missing from the PDF.

When I process the same map and topics to HTML5 output, the <abbreviated-form> elements are processed correctly and this warning is not generated in the log.

I believe this was working in DITA-OT version 3.7, because that's the version bundled with Oxygen 24.1 and the same test files process to PDF correctly to PDF in Oxygen 24.1. It was also working in DITA-OT 2.5.4, which is the one my client currently uses and the <abbreviated-form> elements are rendered correctly to PDF with the correct glossary expansion on first occurrence and the acronym on subsequent occurrences.

I am attaching a test map and topics if you need a test case.
[debug_glossary.zip](https://github.com/dita-ot/dita-ot/files/8855079/debug_glossary.zip)

Cheers,
Gershon

**Comment by raducoravu** [comment, marked as answer]
This bug might be caused by this reported problem:
https://github.com/dita-ot/dita-ot/issues/2773#issuecomment-1005761059
Maybe you can add an extra comment on that issue. It seems to be a side effect of this pull request https://github.com/dita-ot/dita-ot/pull/3210 by @drmacro 

Looking at the dita-ot-3.7.2/plugins/org.dita.pdf2/xsl/fo/abbrev-domain.xsl, it has this line:

      <xsl:variable name="target" select="key('id', substring(@href, 2))[1]" as="element()?"/>

while the same variable in the abbrev-domain.xsl from the Oxygen DITA OT distribution was changed like this:

     <xsl:variable name="target" select="key('id', substring(@href, 2))" as="element()*"/>

**Reply by gershon-pcas** [reply]
Thanks Radu, your suggested fix works. I updated #2773 to indicate that this issue still persists in 3.7.1 and requested a production fix ASAP.

**Reply by raducoravu** [reply]
@gershon-pcas thanks the project is open source, you can also open a pull request if you want to.
https://www.dita-ot.org/contributing

**Reply by chrispy-snps** [reply]
@gershon-pcas - if you are unsure how to create a pull request, I can do this for you.

**Comment by gershon-pcas** [comment]
Hi Chris,

I was going to try to figure out how to do it, but if you can apply the fix I’d be most appreciative.

Cheers,
Gershon

Gershon L Joseph | Senior Information Architect | Precision Content | Phone: +972 (54) 658-3887| TZ: Jerusalem, Israel (GMT+2) | Email: ***@***.******@***.***>| Twitter: @PCASinc<https://twitter.com/PCASInc?lang=en> |  www.precisioncontent.com<http://www.precisioncontent.com/>
 ***@***.***
 Unlock the Knowledge in Your Enterprise™
This email and any files transmitted with it are confidential and intended solely for the use of the individual or entity to whom they are addressed. If you are not the intended recipient you are notified that disclosing, copying, distributing or taking any action in reliance on the contents of this information is strictly prohibited. Please notify us by return email if you have received this email in error. © 2022, Precision Content Authoring Solutions Inc. Toronto, Ontario, Canada


gershon joseph
senior information architect
        [cid:linkedin_32x32_3cde46cf-f061-4d61-8e50-1aae6546c183.png] <https://www.linkedin.com/company/precisioncontent/>  [cid:twitter_32x32_de0e8fcc-5255-43d9-af32-c1dd18c3811f.png] <https://www.twitter.com/pcasinc>      [cid:gptw_5946560a-203c-4c10-9fd3-c413b6647aff.png] <https://www.greatplacetowork.ca/en/certification%E3%85%A4/certified-organizations#certified-organizations/view-sub-list-details/61dde00b19d6d100221fa636/> <https://www.greatplacetowork.ca/en/certification%E3%85%A4/certified-organizations#certified-organizations/view-sub-list-details/61dde00b19d6d100221fa636/>
<https://www.precisioncontent.com/>[cid:precisioncontentlogo1_b18ef66a-3ca1-415b-9c17-f3a1017c9b39.png]<https://www.precisioncontent.com>

180 John St. Toronto, ON Canada M5T 1X5
T: (647) 557-5965<tel:(647)%20557-5965>
***@***.******@***.***>

This email and any files transmitted with it are confidential and intended solely for the use of the individual or entity to whom they are addressed. If you are not the intended recipient you are notified that disclosing, copying, distributing or taking any action in reliance on the contents of this information is strictly prohibited. Please notify us by return email if you have received this email in error. ©2022, Precision Content, Toronto, Ontario, Canada

**Comment by chrispy-snps** [comment]
I looked into why the `[1]` was causing issues, because you would think that the target element would be processed with or without it.

I added some `<xsl:message>` statements to understand what the lookup was returning:

```
    <xsl:message>@href: <xsl:value-of select="@href"/></xsl:message>
    <xsl:message>COUNT: <xsl:value-of select="count(key('id', substring(@href, 2)))"/></xsl:message>
    <xsl:message>ITEM1: <xsl:sequence select="key('id', substring(@href, 2))[1]"/></xsl:message>
    <xsl:message>ITEM2: <xsl:sequence select="key('id', substring(@href, 2))[2]"/></xsl:message>
    <xsl:variable name="target" select="key('id', substring(@href, 2))[1]" as="element()?"/>
```

and here are the results:

```
 [pipeline] @href: #unique_2

 [pipeline] COUNT: 2

 [pipeline] ITEM1: <topicref xmlns:dita-ot="http://dita-ot.sourceforge.net/ns/201007/dita-ot"
 [pipeline]           xmlns:ditaarch="http://dita.oasis-open.org/architecture/2005/"
 [pipeline]           xmlns:opentopic="http://www.idiominc.com/opentopic"
 [pipeline]           xmlns:opentopic-index="http://www.idiominc.com/opentopic/index"
 [pipeline]           xmlns:ot-placeholder="http://suite-sol.com/namespaces/ot-placeholder"
 [pipeline]           class="- map/topicref "
 [pipeline]           first_topic_id="#unique_2"
 [pipeline]           href="#unique_2"
 [pipeline]           id="unique_2"
 [pipeline]           keys="glossary_abs"
 [pipeline]           ohref="5a35493ff2af0f224c071f50c37baad1615f6e05.dita"
 [pipeline]           type="glossentry"
 [pipeline]           xtrc="topicref:2;6:65"
 [pipeline]           xtrf="file:/mnt/c/nobackup/3937/test_map.ditamap">
 [pipeline]    <topicmeta class="- map/topicmeta ">
 [pipeline]       <navtitle class="- topic/navtitle ">Anti-Lock Brakes</navtitle>
 [pipeline]       <?ditaot gentext?>
 [pipeline]       <linktext class="- map/linktext ">Anti-Lock Brakes</linktext>
 [pipeline]    </topicmeta>
 [pipeline] </topicref>

 [pipeline] ITEM2: <glossentry xmlns:dita-ot="http://dita-ot.sourceforge.net/ns/201007/dita-ot"
 [pipeline]             xmlns:ditaarch="http://dita.oasis-open.org/architecture/2005/"
 [pipeline]             xmlns:opentopic-index="http://www.idiominc.com/opentopic/index"
 [pipeline]             class="- topic/topic concept/concept glossentry/glossentry "
 [pipeline]             ditaarch:DITAArchVersion="1.3"
 [pipeline]             domains="(topic concept) (topic concept glossentry) a(props deliveryTarget) (topic equation-d) (topic hazard-d) (topic hi-d) (topic indexing-d) (topic abbrev-d) (topic markup-d) (topic mathml-d) (topic pr-d) (topic relmgmt-d) (topic sw-d) (topic svg-d) (topic ui-d) (topic ut-d) (topic markup-d xml-d)"
 [pipeline]             id="unique_2"
 [pipeline]             oid="anti_lock_brakes"
 [pipeline]             specializations="@props/deliveryTarget"
 [pipeline]             xtrc="glossentry:1;3:35"
 [pipeline]             xtrf="file:/mnt/c/nobackup/3937/anti_lock_brakes.dita">
 [pipeline]    <glossterm class="- topic/title concept/title glossentry/glossterm "
 [pipeline]               xtrc="glossterm:1;4:16"
 [pipeline]               xtrf="file:/mnt/c/nobackup/3937/anti_lock_brakes.dita">Anti-Lock Brakes</glossterm>
 [pipeline]    <glossBody class="- topic/body concept/conbody glossentry/glossBody "
 [pipeline]               xtrc="glossBody:1;5:16"
 [pipeline]               xtrf="file:/mnt/c/nobackup/3937/anti_lock_brakes.dita">
 [pipeline]       <glossSurfaceForm class="- topic/p concept/p glossentry/glossSurfaceForm "
 [pipeline]                         xtrc="glossSurfaceForm:1;6:27"
 [pipeline]                         xtrf="file:/mnt/c/nobackup/3937/anti_lock_brakes.dita">Anti-Lock Brakes (ABS)</glossSurfaceForm>
 [pipeline]       <glossAlt class="- topic/section concept/section glossentry/glossAlt "
 [pipeline]                 id="unique_2_Connect_42_glossAlt_mjb_ply_stb"
 [pipeline]                 xtrc="glossAlt:1;7:45"
 [pipeline]                 xtrf="file:/mnt/c/nobackup/3937/anti_lock_brakes.dita">
 [pipeline]          <glossAcronym class="- topic/title concept/title glossentry/glossAcronym "
 [pipeline]                        xtrc="glossAcronym:1;8:27"
 [pipeline]                        xtrf="file:/mnt/c/nobackup/3937/anti_lock_brakes.dita">ABS</glossAcronym>
 [pipeline]       </glossAlt>
 [pipeline]    </glossBody>
 [pipeline] </glossentry>
```

So the root cause of the issue is that the element lookup by `@id` is returning both a `<topicref>` as well as a `<glossentry>`.

If `[1]` is simply removed, then both elements are passed into the subsequent processing, which happens to work but is not really correct.

Because subsequent processing expects to only process `<glossentry>` elements, a better solution is to filter the lookup accordingly:

```
<xsl:variable name="target" select="key('id', substring(@href, 2))[contains(@class,' glossentry/glossentry ')][1]" as="element()?"/>
```

I will submit a pull request for this.

**Reply by chrispy-snps** [reply]
I submitted pull request #4002 to the `hotfix/3.7.4` branch.

**Reply by chrispy-snps** [reply]
Pull request approved, this fix will be in 3.7.4.

**Comment by gershon-pcas** [comment]
Super!!! Thanks so much for fixing this, Chris. Appreciate it very much.

Cheers,
Gershon

Gershon L Joseph | Senior Information Architect | Precision Content | Phone: +972 (54) 658-3887| TZ: Jerusalem, Israel (GMT+2) | Email: ***@***.******@***.***>| Twitter: @PCASinc<https://twitter.com/PCASInc?lang=en> |  www.precisioncontent.com<http://www.precisioncontent.com/>
 ***@***.***
 Unlock the Knowledge in Your Enterprise™
This email and any files transmitted with it are confidential and intended solely for the use of the individual or entity to whom they are addressed. If you are not the intended recipient you are notified that disclosing, copying, distributing or taking any action in reliance on the contents of this information is strictly prohibited. Please notify us by return email if you have received this email in error. © 2022, Precision Content Authoring Solutions Inc. Toronto, Ontario, Canada


gershon joseph
senior information architect
        [cid:linkedin_32x32_3cde46cf-f061-4d61-8e50-1aae6546c183.png] <https://www.linkedin.com/company/precisioncontent/>  [cid:twitter_32x32_de0e8fcc-5255-43d9-af32-c1dd18c3811f.png] <https://www.twitter.com/pcasinc>      [cid:gptw_5946560a-203c-4c10-9fd3-c413b6647aff.png] <https://www.greatplacetowork.ca/en/certification%E3%85%A4/certified-organizations#certified-organizations/view-sub-list-details/61dde00b19d6d100221fa636/> <https://www.greatplacetowork.ca/en/certification%E3%85%A4/certified-organizations#certified-organizations/view-sub-list-details/61dde00b19d6d100221fa636/>
<https://www.precisioncontent.com/>[cid:precisioncontentlogo1_b18ef66a-3ca1-415b-9c17-f3a1017c9b39.png]<https://www.precisioncontent.com>

180 John St. Toronto, ON Canada M5T 1X5
T: (647) 557-5965<tel:(647)%20557-5965>
***@***.******@***.***>

This email and any files transmitted with it are confidential and intended solely for the use of the individual or entity to whom they are addressed. If you are not the intended recipient you are notified that disclosing, copying, distributing or taking any action in reliance on the contents of this information is strictly prohibited. Please notify us by return email if you have received this email in error. ©2022, Precision Content, Toronto, Ontario, Canada
